File: web/src/lib/sort.test.ts
import { describe, expect, it } from "vitest";
import { sortRecipes } from "./sort";

const a = {
  slug: "a-bouillon",
  title: "Bouillon de châtaigne",
  totalTime: 40,
  updatedAt: "2026-04-10T00:00:00.000Z",
};

const b = {
  slug: "b-flan",
  title: "Flan caramel",
  totalTime: 80,
  updatedAt: "2026-04-20T00:00:00.000Z",
};

const c = {
  slug: "c-katsu",
  title: "Aubergine miso",
  totalTime: 25,
  updatedAt: "2026-04-15T00:00:00.000Z",
};

describe("sortRecipes", () => {
  it("'recent' trie par updatedAt décroissant", () => {
    expect(sortRecipes([a, b, c], "recent")).toEqual([
      "b-flan",
      "c-katsu",
      "a-bouillon",
    ]);
  });

  it("'alpha' trie par titre croissant (collation FR)", () => {
    expect(sortRecipes([a, b, c], "alpha")).toEqual([
      "c-katsu",
      "a-bouillon",
      "b-flan",
    ]);
  });

  it("'duration' trie par totalTime croissant", () => {
    expect(sortRecipes([a, b, c], "duration")).toEqual([
      "c-katsu",
      "a-bouillon",
      "b-flan",
    ]);
  });

  it("'recent' départage les updatedAt identiques par slug", () => {
    const t = "2026-04-28T07:05:10.000Z";
    const x = { ...a, slug: "z-dernier", updatedAt: t };
    const y = { ...b, slug: "a-premier", updatedAt: t };
    const z = { ...c, slug: "m-milieu", updatedAt: t };
    expect(sortRecipes([x, y, z], "recent")).toEqual([
      "a-premier",
      "m-milieu",
      "z-dernier",
    ]);
    expect(sortRecipes([z, x, y], "recent")).toEqual([
      "a-premier",
      "m-milieu",
      "z-dernier",
    ]);
  });

  it("'duration' départage les totalTime identiques par slug", () => {
    const x = { ...a, slug: "z-dernier", totalTime: 30 };
    const y = { ...b, slug: "a-premier", totalTime: 30 };
    expect(sortRecipes([x, y], "duration")).toEqual([
      "a-premier",
      "z-dernier",
    ]);
    expect(sortRecipes([y, x], "duration")).toEqual([
      "a-premier",
      "z-dernier",
    ]);
  });

  it("ne mute pas l'argument d'entrée", () => {
    const arr = [a, b, c];
    sortRecipes(arr, "alpha");
    expect(arr.map((r) => r.slug)).toEqual([
      "a-bouillon",
      "b-flan",
      "c-katsu",
    ]);
  });
});

File: android/app/src/test/java/fr/vferries/cuisine/data/SortTest.kt
package fr.vferries.cuisine.data

import org.junit.Assert.assertEquals
import org.junit.Test

class SortTest {

    private fun meta(
        slug: String,
        title: String,
        totalTime: Int,
        updatedAt: String,
    ) = RecipeMeta(
        slug = slug,
        title = title,
        servings = 2,
        prepTime = 0,
        cookTime = 0,
        totalTime = totalTime,
        difficulty = "moyenne",
        cuisine = "test",
        updatedAt = updatedAt,
    )

    private val a = meta("a", "Bouillon de châtaigne", 40, "2026-04-10T00:00:00.000Z")
    private val b = meta("b", "Flan caramel", 80, "2026-04-20T00:00:00.000Z")
    private val c = meta("c", "Aubergine miso", 25, "2026-04-15T00:00:00.000Z")

    @Test fun recent_sorts_by_updatedAt_descending() {
        assertEquals(
            listOf("b", "c", "a"),
            sortRecipes(listOf(a, b, c), SortMode.RECENT),
        )
    }

    @Test fun recent_breaks_updatedAt_ties_by_slug() {
        val t = "2026-04-28T07:05:10.000Z"
        val x = meta("z-dernier", "X", 10, t)
        val y = meta("a-premier", "Y", 10, t)
        val z = meta("m-milieu", "Z", 10, t)
        val expected = listOf("a-premier", "m-milieu", "z-dernier")
        assertEquals(expected, sortRecipes(listOf(x, y, z), SortMode.RECENT))
        assertEquals(expected, sortRecipes(listOf(z, x, y), SortMode.RECENT))
    }

    @Test fun duration_breaks_totalTime_ties_by_slug() {
        val x = meta("z-dernier", "X", 30, "2026-04-01T00:00:00.000Z")
        val y = meta("a-premier", "Y", 30, "2026-04-02T00:00:00.000Z")
        val expected = listOf("a-premier", "z-dernier")
        assertEquals(expected, sortRecipes(listOf(x, y), SortMode.DURATION))
        assertEquals(expected, sortRecipes(listOf(y, x), SortMode.DURATION))
    }

    @Test fun alpha_sorts_by_title_ascending_french() {
        assertEquals(
            listOf("c", "a", "b"),
            sortRecipes(listOf(a, b, c), SortMode.ALPHA),
        )
    }

    @Test fun duration_sorts_by_totalTime_ascending() {
        assertEquals(
            listOf("c", "a", "b"),
            sortRecipes(listOf(a, b, c), SortMode.DURATION),
        )
    }
}
